discovery/sync_manager.go

```diff
@@ -205,297 +186,301 @@ func (m *SyncManager) syncerHandler() {
+		// a peer and force a historical sync with them.
+		case <-m.cfg.HistoricalSyncTicker.Ticks():
+			m.forceHistoricalSync()
 
 		case <-m.quit:
 			return
 		}
 	}
 }
 
-// queueActiveSyncer queues the given pending active gossip syncer to the end of
-// the round-robin queue.
-func (m *SyncManager) queueActiveSyncer(s *GossipSyncer) {
-	log.Debugf("Queueing next pending active GossipSyncer(%x)",
-		s.cfg.peerPub)
+// createGossipSyncer creates the GossipSyncer for a newly connected peer.
+func (m *SyncManager) createGossipSyncer(peer lnpeer.Peer) *GossipSyncer {
+	nodeID := routing.Vertex(peer.PubKey())
+	log.Infof("Creating new GossipSyncer for peer=%x", nodeID[:])
 
-	delete(m.inactiveSyncers, s.cfg.peerPub)
-	m.pendingActiveSyncers[s.cfg.peerPub] = s
-	m.pendingActiveSyncerQueue.PushBack(s)
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(gossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		peerPub:       nodeID,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		batchSize:     requestBatchSize,
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessageLazy(false, msgs...)
+		},
+	})
+
+	// Gossip syncers are initialized by default in a PassiveSync type
+	// and chansSynced state so that they can reply to any peer queries or
+	// handle any sync transitions.
+	s.setSyncState(chansSynced)
+	s.setSyncType(PassiveSync)
+	return s
 }
 
-// nextPendingActiveSyncer returns the next active syncer pending to be
-// transitioned. If there aren't any, then `nil` is returned.
-func (m *SyncManager) nextPendingActiveSyncer() *GossipSyncer {
-	next := m.pendingActiveSyncerQueue.Front()
-	for next != nil {
-		s := m.pendingActiveSyncerQueue.Remove(next).(*GossipSyncer)
-
-		// If the next pending active syncer is no longer in our lookup
-		// map, then the corresponding peer has disconnected, so we'll
-		// skip them.
-		if _, ok := m.pendingActiveSyncers[s.cfg.peerPub]; !ok {
-			next = m.pendingActiveSyncerQueue.Front()
-			continue
-		}
+// removeGossipSyncer removes all internal references to the disconnected peer's
+// GossipSyncer and stops it. In the event of an active GossipSyncer being
+// disconnected, a passive GossipSyncer, if any, will take its place.
+func (m *SyncManager) removeGossipSyncer(peer routing.Vertex) {
+	m.syncersMu.Lock()
+	defer m.syncersMu.Unlock()
 
-		return s
+	s, ok := m.gossipSyncer(peer)
+	if !ok {
+		return
 	}
 
-	return nil
+	log.Infof("Removing GossipSyncer for peer=%v", peer)
+
+	// We'll stop the GossipSyncer for the disconnected peer in a goroutine
+	// to prevent blocking the SyncManager.
+	go s.Stop()
+
+	// If it's a non-active syncer, then we can just exit now.
+	if _, ok := m.inactiveSyncers[peer]; ok {
+		delete(m.inactiveSyncers, peer)
+		return
+	}
+
+	// Otherwise, we'll need find a new one to replace it, if any.
+	delete(m.activeSyncers, peer)
+	newActiveSyncer := chooseRandomSyncer(
```

Comment:
not used?

Reply:
It is. If you look at `chooseRandomSyncer` you can provide a callback to perform a certain action on each syncer. So by the time this returns, it either returns `nil` because it couldn't perform the action on any syncers, or it returns the first syncer with which we successfully performed the action.

Reply:
Ah, so it is only used to print following debug log?

Reply:
Correct.